### DISCUSSION TITLE
关于相位校正的一些疑问

### DISCUSSION THREAD

**Discussion by HaiMianBBao** [Open3DV/Xema#38] [General]
xema代码里面的相位校正是为了修正反射率突变边界处的相位误差，请问是如何达到去除误差的效果的呢？

**Comment by kimkita** [comment]
现象是在反射率突变的边缘位置会出现点云深度的波动，其中一个原因是由于正弦编码的结构光被相邻的像素干扰所致，在相机成像系统中的点扩散函数正是描述这种“干扰”的，在XEMA程序中将成像系统的点扩散函数假设为了高斯函数，于是通过调节高斯函数的参数，以表示点扩散函数的强度。
在这个程序中，先对采集到的正弦条纹图像进行高斯模糊，然后计算相位对比高斯模糊前的变化程度d，d就作为相位矫正的依据。

特别说明，这种方法针对的是点扩散函数引入的相位误差，另一种误差，例如：当成像的黑白边界成像于图像上某像素中心时，该像素的相位误差理论上不能由这个方式得到正确补偿，其原因是误差来源于像素本身具有面积大小，而非一个理想的点（受限于分辨率）。

**Reply by HaiMianBBao** [reply]
感谢你的回复，像下面这样图的情形
![2024_1_27_10_44](https://github.com/Open3DV/Xema/assets/54971473/eeecbb51-9dfe-4d5a-84df-dde012c9f981)

点云会出现这种尖峰：
![image](https://github.com/Open3DV/Xema/assets/54971473/50108ea5-6ee9-4b8e-be04-4284a3b217ab)

这种情况是不是就没法用代码里面的方式进行补偿了？

**Reply by vic-w** [reply]
这个图看不清楚，能否上传点云的原图？需要压缩成zip文件

**Reply by HaiMianBBao** [reply]
可以上传点云图
[2024-02-26_15-19-55.zip](https://github.com/Open3DV/Xema/files/14402469/2024-02-26_15-19-55.zip)

**Comment by vic-w** [comment]
可以参考代码，理解补偿的原理。
https://github.com/Open3DV/Xema/blob/0866f004ec990731b403685de55c7a8258e64f0d/firmware/management.cu#L1882

这里有一篇论文，也可以参照理解。算法的实现方式虽然跟论文不完全一样，但基本思想是一致的。
[Discontinuity-induced measurement artifact reduction of structured light system based on a deconvolution method.pdf](https://github.com/Open3DV/Xema/files/14402346/Discontinuity-induced.measurement.artifact.reduction.of.structured.light.system.based.on.a.deconvolution.method.pdf)

**Reply by HaiMianBBao** [reply]
论文里面的反卷积思想貌似跟代码里面的思想不太一样

**Reply by vic-w** [reply]
我们的具体做法是：

1. 让本来已经模糊的图像进行一次高斯卷积，变得更模糊了
2. 让这个图像减去原图，得到一个delta
3. 让原图减去这个delta，达到锐化的目的

本质上还是做了一次图像锐化。这个方法应该是叫做Unsharp masking (USM) 。

**Comment by vic-w** [comment]
我再对整个问题做一个完整的陈述。
**这个问题的现象是：**
结构光3D相机（由一个投影仪和一个2D相机组成）在拍摄物体时，在物体表面亮度突变的位置（也就是黑白交界的地方）会出现点云的波动。
- 如果白色区域和2D相机同侧，黑色区域和投影仪同侧，则点云会朝靠近相机的方向偏移
- 如果黑色区域和2D相机同侧，白色区域和投影仪同侧，则点云会朝远离相机的方向偏移

相信大家在拍摄标定板时经常观察到这个现象。
![标定板亮度图](https://github.com/Open3DV/Xema/assets/3446495/4c9e091c-86a5-4b0b-9917-1cd8d06a50ca)
标定板亮度图
![标定板深度图（局部）](https://github.com/Open3DV/Xema/assets/3446495/2ce8cf66-8b16-4f20-893f-e937933c26f5)
标定板深度图（局部）

**产生问题的原因是：**
2D相机镜头成像不清晰。任何镜头都存在光学极限，无法做到完全清晰。这是由于光线在穿过镜头时会发生衍射，导致光点扩散成艾里斑。艾里斑是光学系统成像的最小光斑，其大小与光圈和光线的波长有关。
点扩散函数（PSF）是光学系统对理想点光源的成像响应。PSF的大小和形状决定了光学系统的成像质量。理想情况下，PSF应该是一个无限小的点，但实际上，PSF总是具有一定大小和形状的。
由于PSF的存在，2D相机成像时，相当于对投影条纹图像做了一次高斯模糊（假设模糊的卷积核是高斯函数）。
这个模糊的效果在物体表面亮度均匀的区域并不影响3D成像的效果。这是因为投影条纹是正弦条纹，经过高斯模糊之后，依然保持是正弦条纹，而且每个像素的相位还保持不变。
但如果是黑白交界的区域，正弦条纹就会遭到破坏。假设某一个像素一侧是黑色区域，另一侧是白色区域。那么这个像素成像的亮度，更大程度上受到白色一侧的影响。因为白色一侧反射的能量较多，这个像素被白色的能量所主导。那么解出的相位信息就会偏向白色的一侧。就像是白色一侧把黑色一侧的相位给“吸”过去了。

**解决的方法是：**
用图像锐化的方法抵消PSF产生的影响。
这篇论文里讲述了用反卷积的方法抵消PSF
[Discontinuity-induced measurement artifact reduction of structured light system based on a deconvolution method.pdf](https://github.com/Open3DV/Xema/files/14402346/Discontinuity-induced.measurement.artifact.reduction.of.structured.light.system.based.on.a.deconvolution.method.pdf)
在XEMA的算法实现中，我们用的是Unsharp masking (USM) 方法来实现图像的锐化。
具体做法是：
1. 让本来已经模糊的图像进行一次高斯卷积，变得更模糊了
2. 让这个图像减去原图，得到一个delta
3. 让原图减去这个delta，达到锐化的目的

这个方法并不是自适应的，需要根据每次拍摄的情况调整参数（高斯卷积的标准差）。
![ede212a71e914368d3f04ad0e7ad082](https://github.com/Open3DV/Xema/assets/3446495/f5c8d7b8-23f8-4e6c-833b-9d24772f7bb8)
当调整到合适的参数时，黑白边界处的点云得到了矫正，深度图上只剩下一些随机噪声。
![矫正前](https://github.com/Open3DV/Xema/assets/3446495/9c5da777-2e0b-472a-bc12-b3f969109311)
矫正前
![矫正后](https://github.com/Open3DV/Xema/assets/3446495/45397aad-2f5a-485e-b809-f54eb2b18171)
矫正后

**Comment by zhouscd** [comment]
研究了一下代码，发现算法里的步骤是这样的：
1、是对2D图像进行高斯卷积使其更模糊
2、对模糊的图像进行相位计算
3、原图也进行相位计算
4、最终相位 = 原图相位 - 模糊图像相位+原图相位

这种做法跟直接对2D图像进行Unsharp masking后再计算相位有什么区别或者优势吗？

**Reply by vic-w** [reply]
对原图模糊，可以尽量靠前抑制噪声。其余的运算推迟到相位图，运算量小了。

**Reply by zhouscd** [reply]
相位计算的计算量也不小吧？

**Reply by vic-w** [reply]
原始图有多张，相位图只有一张。说优势也算不上。
